libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -0,0 +1,1221 @@
+		else
+		{
+			// parse as rational number
+			std::optional<rational> tmp = parseRational(valueString);
+			if (!tmp)
+				return std::nullopt;
+			value = *tmp;
+		}
+	}
+	catch (...)
+	{
+		return std::nullopt;
+	}
+	switch (_literal.subDenomination())
+	{
+	case Literal::SubDenomination::None:
+	case Literal::SubDenomination::Wei:
+	case Literal::SubDenomination::Second:
+		break;
+	case Literal::SubDenomination::Gwei:
+		value *= bigint("1000000000");
+		break;
+	case Literal::SubDenomination::Ether:
+		value *= bigint("1000000000000000000");
+		break;
+	case Literal::SubDenomination::Minute:
+		value *= bigint("60");
+		break;
+	case Literal::SubDenomination::Hour:
+		value *= bigint("3600");
+		break;
+	case Literal::SubDenomination::Day:
+		value *= bigint("86400");
+		break;
+	case Literal::SubDenomination::Week:
+		value *= bigint("604800");
+		break;
+	case Literal::SubDenomination::Year:
+		value *= bigint("31536000");
+		break;
+	}
+
+	return value;
+}
+}
+
+bool TypeInference::visit(Literal const& _literal)
+{
+	auto& literalAnnotation = annotation(_literal);
+	if (_literal.token() != Token::Number)
+	{
+		m_errorReporter.typeError(4316_error, _literal.location(), "Only number literals are supported.");
+		return false;
+	}
+	std::optional<rational> value = rationalValue(_literal);
+	if (!value)
+	{
+		m_errorReporter.typeError(6739_error, _literal.location(), "Invalid number literals.");
+		return false;
+	}
+	if (value->denominator() != 1)
+	{
+		m_errorReporter.typeError(2345_error, _literal.location(), "Only integers are supported.");
+		return false;
+	}
+	literalAnnotation.type = m_typeSystem.freshTypeVariable(Sort{{annotation().builtinClasses.at(BuiltinClass::Integer)}});
+	return false;
+}
+
+
+namespace
+{
+// TODO: put at a nice place to deduplicate.
+TypeRegistration::TypeClassInstantiations const& typeClassInstantiations(Analysis const& _analysis, TypeClass _class)
+{
+	auto const* typeClassDeclaration = _analysis.typeSystem().typeClassDeclaration(_class);
+	if (typeClassDeclaration)
+		return _analysis.annotation<TypeRegistration>(*typeClassDeclaration).instantiations;
+	// TODO: better mechanism than fetching by name.
+	auto const& annotation = _analysis.annotation<TypeRegistration>();
+	auto const& inferenceAnnotation = _analysis.annotation<TypeInference>();
+	return annotation.builtinClassInstantiations.at(
+		inferenceAnnotation.builtinClassesByName.at(
+			_analysis.typeSystem().typeClassName(_class)
+		)
+	);
+}
+}
+
+void TypeInference::unifyGeneralized(Type _type, Type _scheme, std::vector<Type> _monomorphicTypes, langutil::SourceLocation _location)
+{
+	solUnimplementedAssert(_monomorphicTypes.empty(), "unsupported");
+	unify(_type, m_env->fresh(_scheme), _location);
+}
+
+experimental::Type TypeInference::polymorphicInstance(Type _scheme, langutil::SourceLocation _location)
+{
+	TypeVariable result = m_typeSystem.freshTypeVariable({});
+	unifyGeneralized(result, _scheme, {}, _location);
+	return result;
+}
```

Comment:
Why do we introduce a fresh variable as a middleman here? Wouldn't the behavior be the same if we just returned the result of `fresh()` from `polymorphicInstance()`? Other than the fact that it would not need `resolve()`, but you said yourself that it's just an implementation detail.

Reply:
Addressed by #14635.